Add hdfs read support

Arty-Maly opened the pull request:
Add Hdfs read support. The design is similar to the s3 connector. 
There is an HdfsFileSystem class that creates the client and vends HdfsReadFiles. 

HdfsReadFile is in charge of returning data from the file via the pread methods. On each pread HdfsReadFile creates an hdfsFile and reads the requested data, after the read is finished the file is closed. Since each pread creates its own file object, the pread methods are thread safe.

To test this I depend on the hadoop executable that has a local mini cluster implementation. For my tests I start a mini cluster and read from it.

pedroerp commented on scripts/setup-circleci.sh:
```diff
@@ -65,9 +65,12 @@ wget_and_untar https://boostorg.jfrog.io/artifactory/main/release/1.72.0/source/
 
 wait  # For cmake and source downloads to complete.
 
+cp -a hadoop-2.10.1 /usr/local/
```
Don't we need some logic below to build and properly install it?

Arty-Maly replied:
No, this is already prebuilt so we just need to copy it onto our path.

pedroerp replied:
Pre-built for which architecture? Is it something that's supposed to work across platforms and architectures? If not, it would be a good idea to get the sources and compile, like we do with the other dependencies.

frankobe replied:
@pedroerp This contains jar dependencies so it will work across platform as long as correct java version is installed

pedroerp replied:
ah, this is a java library! Never mind then :) How do we communicate with it, btw? Is there a JNI wrapper somewhere?

Arty-Maly replied:
I use HdfsMinicluster class to issue commands to it. I only need the mini cluster implementation from these libraries for testing the hdfs integration. 
The minicluster is essentially a local node made for testing out functionality.

pedroerp commented on scripts/setup-ubuntu.sh:
```diff
@@ -122,9 +122,16 @@ function install_folly {
 }
 
+function install_libhdfs3 {
+  github_checkout apache/hawq master
```
is libhdfs only part of the hawq project? is it not provided as a standalone library?

Arty-Maly replied:
I did find this: https://github.com/erikmuttersbach/libhdfs3 but it seems to me as not as active as the hawq repo. 

Also from this ticket it looks like for now apache/hawq is where libhdfs3 is being maintained.
https://issues.apache.org/jira/browse/HDFS-6994

pedroerp commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,99 @@
+/*
```
@majetideepak since you were the one creating this interface some time ago, could you help reviewing this PR?

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.h:
```diff
@@ -0,0 +1,55 @@
+ * limitations under the License.
+ */
+#include <hdfs/hdfs.h>
+#include "HdfsReadFile.h"
```
Is this header required here? Sufficient to put it in the `.cpp` file?

Arty-Maly replied:
It is not. I removed it.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.h:
```diff
@@ -0,0 +1,55 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <hdfs/hdfs.h>
```
We used the PIMPL idiom for S3FileSystem to avoid including a third-party header in this header. PIMPL helps keep the API clean from third-party headers.

Arty-Maly replied:
Understood. Will do.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.h:
```diff
@@ -0,0 +1,55 @@
+
+class HdfsFileSystem : public FileSystem {
+ private:
+  hdfsFS hdfsClient_;
```
Should this be a `std::shared_ptr`?

Arty-Maly replied:
hdfsFS is already a pointer.  
The hdfs library defines it like this: `typedef struct HdfsFileSystemInternalWrapper * hdfsFS;`
I dont think we should wrap it again in a shared pointer.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.h:
```diff
@@ -0,0 +1,45 @@
+
+class HdfsReadFile final : public ReadFile {
+ public:
+  explicit HdfsReadFile(hdfsFS hdfs, std::string_view path);
```
use `const` for the `path`.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.h:
```diff
@@ -0,0 +1,45 @@
+  hdfsFileInfo* fileInfo_;
+  std::string filePath_;
+};
+} // namespace facebook::velox
```
nit: new line here.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.cpp:
```diff
@@ -0,0 +1,103 @@
+void HdfsReadFile::preadInternal(uint64_t offset, uint64_t length, char* pos)
+    const {
+  checkFileReadParameters(offset, length);
+  auto file = hdfsOpenFile(hdfsClient_, filePath_.data(), O_RDONLY, 0, 0, 0);
```
Opening the file for every `read` is not lightweight correct?. If I recall, the library has to talk to get NameNode and get a list of block locations which then are used to open the file. We need to figure out a way to avoid this.

Arty-Maly replied:
`hdfsOpenFile` returns a struct with a input/output stream. 

The block location logic actually happens on `hdfsRead` 

Inside `hdfsRead` there is logic to locate blocks:

1)Get block information from namenode
2)Check if we reach the end of file
3)seek to the right block to read
4) setup block reader

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.cpp:
```diff
@@ -0,0 +1,103 @@
+      startPoint,
+      endPoint)
+}
+} // namespace facebook::velox
```
nit: new line.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.cpp:
```diff
@@ -0,0 +1,103 @@
+  return false;
+}
+
+void HdfsReadFile::checkFileReadParameters(uint64_t startPoint, uint64_t length)
```
is `offset` better instead of `startPoint`?

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/tests/HdfsMiniCluster.cpp:
```diff
@@ -0,0 +1,105 @@
+ * limitations under the License.
+ */
+
+#include "HdfsMiniCluster.h"
```
space after include.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/tests/HdfsFileSystemTest.cpp:
```diff
@@ -0,0 +1,316 @@
+#include "HdfsMiniCluster.h"
+#include "gtest/gtest.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/exec/tests/utils/TempFilePath.h"
```
club all Velox headers together on top?

Arty-Maly replied:
Im not sure what you are suggesting. Do you want me to put all the "velox" dependencies above the other ones?

so like this:

#include "velox/common/file/FileSystems.h"
#include "velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.h"
#include "velox/exec/tests/utils/TempFilePath.h"
#include <boost/format.hpp>
#include <folly/portability/GMock.h>
...

majetideepak replied:
Since you already made this change, this is fine. I noticed that the order of includes is inconsistent in a few other files as well. I see here (https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#cpp-style) that clang-format is supposed to handle this, but it isn't currently.

Arty-Maly replied:
Thanks for the link. I will review that again to fix other discrepancies

majetideepak reviewed (commented):
Thanks for refactoring to align with S3FileSystem. I left some comments. I am yet to look at the tests.
My major concern is opening the HDFS file for every `preadInternal` call.
I see that the motivation is to make the `preadInternal` call thread-safe.
@frankobe and I discussed an alternative of not allowing the `fileHandles` to be cached which is the source of thread safety issue I believe. What happened to that idea?

Arty-Maly commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,99 @@
+HdfsServiceEndpoint HdfsFileSystem::getServiceEndpoint(
+    const std::shared_ptr<const Config>& config) {
+  auto configuration = config.get();
+  VELOX_CHECK(
```
No, I assumed it would be populated in the config. So however the config is constructed, we would have to add the new keys.

Arty-Maly commented:
> Thanks for refactoring to align with S3FileSystem. I left some comments. I am yet to look at the tests. My major concern is opening the HDFS file for every `preadInternal` call. I see that the motivation is to make the `preadInternal` call thread-safe. @frankobe and I discussed an alternative of not allowing the `fileHandles` to be cached which is the source of thread safety issue I believe. What happened to that idea?

@majetideepak 

I also had the same concerns and was hoping to discuss this. 
I was influenced by the S3FileSystem design, to me it seemed on each read you would pay a similar price: 
1)Create Request  2) Call S3 backend  3) Get Result;
I was thinking (2) will be pretty expensive as well. 

I did some testing with locking on `preadInternal` vs opening file on each read. In Velox threads share the ReadFile objects, so two threads can get a task to read different positions with the same read file. 
For testing I used TPCH queries. 
Performance wise locking and opening on each `preadInternal` is pretty much the same. 
With locks you get threads that use the same ReadFile object waiting to acquire the lock.
With opening a file on each read your network throughput might suffer since now all the threads can proceed at the same time. 

I also tested out setting the cache to zero which did not allow the file handles to be stored. 
Setting cache to zero is equivalent with opening a new file on each read, the cache would create a new object on each file handle request. But as of now thats a global setting which will affect other file systems as well. (So if you were to read from hdfs and then from s3 the caching behaviour would be the same.)

I think relying on a configuration for thread safety is not a good strategy. If we would pursue this path I would then change `preadInternal` to use locks just to guarantee thread safety even though the objects will all be different.

majetideepak commented:
> Setting cache to zero is equivalent with opening a new file on each read, the cache would create a new object on each file handle request. 

The `file handle` object is created only once for each split and not per read? `asyncCache` creates another for the `CachedBufferedInputFactory`.
```
void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
...
  fileHandle_ = fileHandleFactory_->generate(split_->filePath);
```

> But as of now thats a global setting which will affect other file systems as well. (So if you were to read from hdfs and then from s3 the caching behaviour would be the same.)

I was thinking we could add a field to `ReadFile` that tells the `FileHandle` object it is not thread-safe and cannot be cached. That way, the FileHandleFactory can always return a new `FileHandle` and this will be per `split` and not per `read`.

But based on your other comment, I don't think this refactor is going to help if every read call is getting the block locations.
I am surprised  `libhdfs3` has no means to cache block locations at least when reading from the same HDFS block.
> The block location logic actually happens on hdfsRead

Arty-Maly commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,99 @@
+
+std::function<bool(std::string_view)> HdfsFileSystem::schemeMatcher() {
+  return [](std::string_view filename) {
+    return filename.find('/') == 0 || filename.find(scheme) == 0;
```
you are right, will fix

Arty-Maly commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.h:
```diff
@@ -0,0 +1,55 @@
+
+  void remove(std::string_view path) override;
+
+  static std::function<bool(std::string_view)> schemeMatcher();
```
So I need it to be a std::function<bool> since I use it with `registerFileSystem` which takes two functions: 
a scheme mathcer and a filesystem generator.

I can change the name of this function to `isHdfsFile` if that would make it clearer.

Arty-Maly commented:
>I was thinking we could add a field to ReadFile that tells the FileHandle object it is not thread-safe and cannot be cached. That way, the FileHandleFactory can always return a new FileHandle and this will be per split and not per read.

I thought `pread` as an os file method is thread safe so thats why I pushed to make these `pread` functions thread safe. I am wary of adding additional logic in the factory level to make it thread safe. I am thinking of future maintainability when someone else would add features not knowing that the hdfs read has to be behind the FileHandleFactory thread safe logic. 

However I am in no means an expert here so I will probably give more weight to opinions from the community. 

>But based on your other comment, I don't think this refactor is going to help if every read call is getting the block locations.
I am surprised libhdfs3 has no means to cache block locations at least when reading from the same HDFS block.

You are correct, all of the expensive operations are in the read file method.

majetideepak commented:
> You are correct, all of the expensive operations are in the read file method.

@Arty-Maly Do you know if the Java library has the same overhead? Or is this a limitation of the libhdfs3 library?

majetideepak reviewed (commented):
@Arty-Maly, thanks for addressing the review comments. Made a few more comments.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/tests/HdfsMiniCluster.h:
```diff
@@ -0,0 +1,63 @@
+  void setupEnvironment(const std::string& homeDirectory);
+
+  std::shared_ptr<::boost::process::child> serverProcess_;
+  boost::filesystem::path exePath;
```
`exePath_` and `env_` below to be in line with the coding guidelines.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  static HdfsServiceEndpoint getServiceEndpoint(const Config* config) {
+    auto kHdfsHost = config->get("hdfs_host");
```
use camelCase for local variables https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#naming-conventions

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,116 @@
+    VELOX_CHECK(
+        kHdfsHost.hasValue(),
+        "kHdfsHost is empty, configuration missing for hdfs host");
+    auto kHdfsPort = config->get("hdfs_port");
```
use camelCase

Arty-Maly commented on scripts/setup-macos.sh:
```diff
@@ -145,6 +145,12 @@ function install_re2 {
   cmake_install -DRE2_BUILD_TESTING=OFF
 }
 
+function install_libhdfs3 {
```
The reason I left this in setup mac os is because setup-adapters.sh has linix specific commands, for example: it uses yum on line 66

majetideepak replied:
We ideally do not want users to build and install libhdfs3 by default.
Can you guard line 66 in `setup-adapters.sh` using a check for the OS type for now? I will look into fixing that in another PR. 
```
if [[ "$OSTYPE" == "linux-gnu"* ]]; then
```

Arty-Maly replied:
Yes no problem.

Arty-Maly commented on scripts/setup-adapters.sh:
```diff
@@ -13,6 +13,9 @@
 # limitations under the License.
 
+SCRIPTDIR=$(dirname "$0")
+source $SCRIPTDIR/helper-functions.sh
```
I wanted to re-use the github-checkout and cmake_install functions. Currently we copy that logic everywhere in our scripts

majetideepak replied:
is this file checked in?

Arty-Maly replied:
Thats what I just realized, its not checked in. Updating now with the comment changes.

Arty-Maly commented:
> > You are correct, all of the expensive operations are in the read file method.
> 
> @Arty-Maly Do you know if the Java library has the same overhead? Or is this a limitation of the libhdfs3 library?

I am not sure, I will double check today and get back to you.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/tests/HdfsFileSystemTest.cpp:
```diff
@@ -0,0 +1,331 @@
+  }
+}
+
+TEST_F(HdfsFileSystemTest, multipleThreadsWithReadFile) {
```
Isn't  `multipleThreadsWithReadFile` redundant given `multipleThreadsWithFileSystem`

majetideepak replied:
Got it. Thanks.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/tests/HdfsFileSystemTest.cpp:
```diff
@@ -0,0 +1,331 @@
+  hdfsBuilderSetNameNodePort(builder, 7878);
+  auto hdfs = hdfsBuilderConnect(builder);
+  HdfsReadFile readFile(hdfs, destinationPath);
+  verifyFailures(&readFile);
```
should we inline `verifyFailures` since it's being used only once?

Arty-Maly replied:
I think abstracting into a method is ok, there is a lot going on there. And putting it all in the test method would make it harder to understand whats going on. Plus there are still some methods that are not supported by the hdfs adapter so I think laying out a structure for the future test case writer is beneficial.

majetideepak replied:
sounds good.

majetideepak commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,115 @@
+    auto builder = hdfsNewBuilder();
+    hdfsBuilderSetNameNode(builder, endpointInfo.host.c_str());
+    hdfsBuilderSetNameNodePort(builder, endpointInfo.port);
+    hdfsClient_ = hdfsBuilderConnect(builder);
```
Is there a default timeout for this if the hdfs cluster is down? Is this configurable in the library?
Can you add a test to read when the hdfs cluster is down?

Arty-Maly replied:
You can configure the settings 2 ways. Via the builder construction (passing the seetings in) or via hdfs-client.xml.
The default is to look for that file in the current directory or by `LIBHDFS3_CONF` env variable.

Just so I have it correct, the test you are suggesting is:

1) hdfsBuilderConnect()
2) kill the test cluster
3) hdfsRead()
4) verify correct behaviour

majetideepak replied:
1) We must add the test you suggested. 
2) We must also add another test where we create an HDFSFileSystem (i.e invoke hdfsBuilderConnect()) when the hdfs cluster is down.

Arty-Maly replied:
Ok I agree. Good catch.

majetideepak replied:
Can you also write a comment in `HdfsFileSystem.cpp` about how to configure libhdfs3 as you described in your comment above. Thanks.

majetideepak replied:
The header `HdfsFileSystem.h` is probably a better place for the comment.

Arty-Maly replied:
Will do

majetideepak reviewed (commented):
@Arty-Maly These should be my final review comments. The remaining changes look good. Thanks for your patience.

zhouyuan commented on velox/connectors/hive/storage_adapters/hdfs/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+add_executable(velox_hdfs_file_test HdfsFileSystemTest.cpp HdfsMinicluster.cpp)
```
```suggestion
add_executable(velox_hdfs_file_test HdfsFileSystemTest.cpp HdfsMiniCluster.cpp)
```
a typo in the file name

Arty-Maly commented on scripts/helper-functions.sh:
```diff
@@ -0,0 +1,56 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function github_checkout {
```
You mean add this as a separate PR? Or put up a separate PR which then uses these functions in the mac os and ubuntu setup scripts? 

I was thinking I can add this script with this PR and only use it with my changes, and as a follow up clean up the mac os and ubuntu scripts.

majetideepak reviewed (commented):
@Arty-Maly Just FYI that another PR that adds support for GCP https://github.com/facebookincubator/velox/pull/1565 was opened yesterday.

majetideepak commented on scripts/setup-adapters.sh:
```diff
@@ -13,6 +13,9 @@
 # limitations under the License.
 
+SCRIPTDIR=$(dirname "$0")
+source $SCRIPTDIR/helper-functions.sh
```
shouldn't this be `setup-helper-functions.sh`? And how do the tests pass without this?

Arty-Maly replied:
Ok makes sense to me.

Arty-Maly commented on scripts/helper-functions.sh:
```diff
@@ -0,0 +1,56 @@
+    -DBUILD_TESTING=OFF \
+    "$@"
+  ninja -C "${BINARY_DIR}" install
+}
```
Got it.
As a side note, are new lines expected in every file? For example if I run cmake-format sometimes it removes new lines at the end in the C++ files. 
So since cmake-format was removing them,  I don't add them

majetideepak replied:
Ideally, add a new line. If cmake-format removes it, then leave it at that.

majetideepak commented:
@mbasmanova, @pedroerp This PR looks ready to me. Can you make a pass? There is one step remaining which is to enable HDFS testing in the circleci config.

majetideepak commented on .circleci/config.yml:
```diff
@@ -316,7 +316,7 @@ jobs:
             - .ccache/
 
-  linux-parquet-s3-build:
+  linux-parquet-hive-storage-adapters-build:
```
@Arty-Maly you have to rename this under `jobs:` on line 21 as well. Looks like CI jobs did not run at all.

Arty-Maly replied:
Is there some file I am missing? I also was confused why there is no circle ci step. I assume it should at least fail?

majetideepak replied:
I am surprised as well.
@pedroerp, @mbasmanova can someone with admin access to CircleCI help see why this PR is not invoking CircleCI runs?

majetideepak replied:
I don't see any runs here as well https://app.circleci.com/pipelines/github/facebookincubator/velox?filter=all

mbasmanova replied:
Could that be due to changes to .circleci/config.yml ?

majetideepak replied:
The changes to `.circleci/config.yml` look correct and are required for this PR. Is there a way we can get more information?

Arty-Maly replied:
I did not get any runs in a prior revision before touching .circleci/config.yml

mbasmanova commented on scripts/setup-helper-functions.sh:
```diff
@@ -0,0 +1,57 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function github_checkout {
```
Nice refactor. Perhaps, submit this as a separate PR. This would make it easier to review. I'd be curious if that PR will trigger CircleCI runs or not.

mbasmanova commented:
@majetideepak Deepak, would you like to duplicate that PR to see if your PR will get CI runs?

mbasmanova commented:
@kgpai Krishna, would you know why CI doesn't run on this PR?

kgpai commented:
Hi @Arty-Maly , I have not seen a run for this pull request in circeci for any of your commits. Typically that means the config is broken or wrong, but I tried a yaml linter + downloaded and verified it locally and looks fine to me. 
For the purposes of an experiment, can you just revert your .config/circleci change and push. If the circelCI job starts then it points to the config being misconfigured somehow , if not well we need to investigate more.

Arty-Maly commented:
> Hi @Arty-Maly , I have not seen a run for this pull request in circeci for any of your commits. Typically that means the config is broken or wrong, but I tried a yaml linter + downloaded and verified it locally and looks fine to me. For the purposes of an experiment, can you just revert your .config/circleci change and push. If the circelCI job starts then it points to the config being misconfigured somehow , if not well we need to investigate more.

I will revert it, however initially I did not have any changes to `.config/circleci` and still did not trigger circleci

majetideepak commented:
I see the runs now after I opened a new PR here https://github.com/facebookincubator/velox/pull/1585. Correlation without causation?

mbasmanova commented:
> I see the runs now after I opened a new PR here #1585. Correlation is not causation?

Ha-ha. Happy to see CI running on this PR. Thank you, Deepak!

Arty-Maly commented:
Weirdly I also opened a pr but dont have circleci https://github.com/facebookincubator/velox/pull/1587

Arty-Maly commented:
And now circle ci is removed from this PR too..

Arty-Maly commented:
> And now circle ci is removed from this PR too..

@majetideepak @mbasmanova

kgpai commented:
Hi All, 
I am going to open a ticket with CircleCI support to help investigate. Will let you know the findings.

kgpai commented:
Heres the ticket : https://support.circleci.com/hc/en-us/requests/109653 .

kgpai commented:
@Arty-Maly , There can be 2 possible causes : 
 1. CircleCI support says 
  "checked our internal logs and found the following error message from GitHub: Received 401 fetching config file due to "Bad credentials".

If this issue was caused by stale permissions, it should be fixed by deleting the user cache. I went ahead and did that. Can you please have the user open a new PR and let me know how it goes?" 
2. Meta OSS Support is also asking to ensure you dont have CircleCI enabled on your personal fork. 

For 1. unfortunately, maybe make a new branch and resubmit the same PR - we might loose some comment history (we can link this PR from that) , but I reckon that is ok.

Arty-Maly commented:
Lets merge this pr first: https://github.com/facebookincubator/velox/pull/1427 (setup script changes)

Then I will rebase those changes and incorporate them here.

majetideepak commented:
@kgpai I noticed here https://github.com/facebookincubator/velox/pull/1587 that closing and re-opening the PR caused CI to trigger runs. It stays that way even if @Arty-Maly pushes new commits. Seems to work here so far as well.

kgpai commented:
@majetideepak I will let CCI know about this, might be a bug on their end.

Arty-Maly commented on scripts/setup-adapters.sh:
```diff
@@ -28,10 +31,33 @@ function install_aws-sdk-cpp {
   cmake_install -DCMAKE_BUILD_TYPE=Debug -DBUILD_SHARED_LIBS:BOOL=OFF -DMINIMIZE_SIZE:BOOL=ON -DENABLE_TESTING:BOOL=OFF -DBUILD_ONLY:STRING="s3;identity-management" -DCMAKE_INSTALL_PREFIX="${DEPENDENCY_DIR}/install"
 }
 
+function install_libhdfs3 {
```
So I have an error here in the circle ci image. 
```
Find GoogleTest include path: /root/adapter-deps/hawq/depends/thirdparty/googletest/googletest/include
-- Find GoogleMock include path: /root/adapter-deps/hawq/depends/thirdparty/googletest/googlemock/include
-- Find Gtest library path: /root/adapter-deps/hawq/depends/thirdparty/googletest/build/googlemock/gtest/libgtest.a
-- Find Gmock library path: /root/adapter-deps/hawq/depends/thirdparty/googletest/build/googlemock/libgmock.a
-- Performing Test GoogleTest_CHECK_FINE
-- Performing Test GoogleTest_CHECK_FINE - Failed
```


It finds gtest but then fails on the check. I have downloaded the circle ci image and ran it in docker to reproduce it but still can't figure out why its failing. I will take another look at it tomorrow. 
As a sanity I even do a system wide gtest install on L51

kgpai commented:
FYI , Response from CCI: 

```
we were seeing 401 errors in our logs for the user Arty-Maly. This can happen when the user has previously authenticated with CircleCI, but has since revoked that authentication.

This can cause an issue on our end, and the PR will fail to trigger any builds due to the above. We are working on showing an error in the pipelines UI to alert users to why it is happening, and is being looked at this quarter.

To resolve this, I have removed the cached authentication for the user on our end. This should allow any future PRs from this user to trigger builds. I believe also pushing a commit to the branch with the open PR should also trigger builds (for PRs already created).
```

I will close the ticket as ci is working now.

Arty-Maly commented on .circleci/config.yml:
```diff
@@ -346,14 +346,25 @@ jobs:
+            tar xf hadoop-2.10.1.tar.gz
+            chmod +x hadoop-2.10.1
+            cp -a hadoop-2.10.1 /usr/local/
+            yum install java-1.8.0-openjdk
```
Should I put installing java as part of the circle ci image instead? into the setup circleci script.
If I do that how do I generate a new image?

Arty-Maly commented:
@majetideepak I think Ill have to update the circle ci docker image.  I need to set JAVA_HOME since hadoop minicluster uses that. 
I dont think I can set it through the config yaml. How do I generate the new image? I added JAVA_HOME to the dockfile and also added java 8 installation to setup circlci.

Arty-Maly commented:
@majetideepak @mbasmanova I have fixed the failure that was causing circle ci to fail. Can you take another look please?

majetideepak commented:
> I think Ill have to update the circle ci docker image.

@Arty-Maly is this still a requirement?

majetideepak commented:
>  I have fixed the failure that was causing circle ci to fail. Can you take another look please?

@Arty-Maly Great! I will take a look today. Thanks.

Arty-Maly commented:
> 

I added my setup to the circle ci  yml. But that means we will download java everytime. If that is ok, then it is solved.

majetideepak commented:
> > 
> 
> I added my setup to the circle ci yml. But that means we will download java everytime. If that is ok, then it is solved.

It is okay for now. I will file an issue after this lands to bake the S3 and HDFS test dependencies into the CI image.

majetideepak reviewed (commented):
@Arty-Maly The code changes look good. However, I noticed a lot of logging here related to hdfs test. Can we supress some of this?
```
127: 22/05/17 05:21:43 INFO impl.FsDatasetImpl: Adding block pool BP-973011278-172.25.0.3-1652764901824
127: 22/05/17 05:21:43 INFO impl.FsDatasetImpl: Scanning block pool BP-973011278-172.25.0.3-1652764901824 on volume /root/project/_build/release/velox/connectors/hive/storage_adapters/hdfs/tests/target/test/data/dfs/data/data1/current...
127: 22/05/17 05:21:43 INFO impl.FsDatasetImpl: Scanning block pool BP-973011278-172.25.0.3-1652764901824 on volume /root/project/_build/release/velox/connectors/hive/storage_adapters/hdfs/tests/target/test/data/dfs/data/data2/current...
127: 22/05/17 05:21:43 INFO hdfs.MiniDFSCluster: dnInfo.length != numDataNodes
127: 22/05/17 05:21:43 INFO hdfs.MiniDFSCluster: Waiting for cluster to become active
127: 22/05/17 05:21:43 INFO impl.FsDatasetImpl: Time taken to scan block pool BP-973011278-172.25.0.3-1652764901824 on /root/project/_build/release/velox/connectors/hive/storage_adapters/hdfs/tests/target/test/data/dfs/data/data1/current: 23ms
```
```
ReadShortCircuitInfo.cpp: 220: HdfsIOException: ReadShortCircuitInfoBuilder: "dfs.domain.socket.path" is not set
```
https://circleci.com/api/v1.1/project/github/facebookincubator/velox/45099/output/111/0?file=true&allocation-id=62832b92aacc2a5491a3cbce-0-build%2F1FA20C3

majetideepak commented:
Compared with the main's log, this test added about 13K more log lines.
https://app.circleci.com/pipelines/github/facebookincubator/velox/8139/workflows/29d21ea0-c944-41e8-8550-0bc97e3beb91/jobs/45213

Arty-Maly commented:
@majetideepak I  the log spam, I added hdfs-client.xml config to .circle folder and also restricted the hdfs mini cluster log spam via an env variable.

frankobe reviewed (approved):
LGTM! I will add a followup pr to enable HDFS in presto-cpp

majetideepak reviewed (approved):
+1 thanks @Arty-Maly

majetideepak commented:
@mbasmanova this PR is ready.

biswapesh commented:
Can we not cache HDFS file handles in a thread safe manner? Why is that?

majetideepak commented:
@biswapesh The current implementation in this PR does enable caching HDFS file handles.

zhouyuan commented on velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:
```diff
@@ -0,0 +1,113 @@
+std::unique_ptr<ReadFile> HdfsFileSystem::openFileForRead(
+    std::string_view path) {
+  if (path.find(kScheme) == 0) {
+    path.remove_prefix(kScheme.length());
```
@Arty-Maly not familiar with presto but usually for spark/hadoop based workload, the URI in each split would be like `hdfs://host:port/hdfs_file.parquet`, 
The code only removed scheme, is there a way to check and remove the `host:port` ?

Arty-Maly replied:
Yes, will remove

Arty-Maly replied:
@zhouyuan  I removed the `host:port`

Arty-Maly commented:
@majetideepak  is there anything else required in this PR?

majetideepak commented:
@Arty-Maly The changes look good. In a follow-up PR, we should support `hdfs://host:port/hdfs_file.parquet` URI where the host and port are derived from the URI instead of the config.

@pedroerp @mbasmanova can someone merge this?

Arty-Maly commented:
@majetideepak 
I also added cutting out `host:port` from the file path: `hdfs://host:port/hdfs_file.parquet` as @zhouyuan pointed out. 

We have different logic internally so I was not aware that paths are passed as `hdfs://host:port/hdfs_file.parquet`

That is now fixed.

frankobe commented:
@pedroerp can u help land this pr, thx!

pedroerp commented:
I got caught up fixing our internal dashboard so we don't miss these sort of things in the future again. Working on importing it now.

kevinwilfong commented:
Could you rebase, it looks like .circleci/config.yml is out of date

pedroerp commented:
@Arty-Maly @frankobe could you guys help rebasing it?

Arty-Maly commented:
@pedroerp rebase is done.

shiyu-bytedance commented on .circleci/config.yml:
```diff
@@ -18,7 +18,7 @@ workflows:
     jobs:
       - linux-build
       - linux-build-options
-      - linux-parquet-s3-build
```
@majetideepak Is linux-parquet-hive-storage-adapters-build replacing remove linux-parquet-s3? Why remove it?

majetideepak replied:
@shiyu-bytedance this is simply a rename needed to accommodate hdfs testing.
